sympy/sets/handlers/union.py

```diff
@@ -138,6 +142,68 @@ def union_sets(a, b): # noqa:F811
             FiniteSet(*[x for x in a if b.contains(x) != True]), b))
     return None
 
+@dispatch(ImageSet, ImageSet)  # type: ignore
+def union_sets(a, b): # noqa:F811
+    """Simplify union of two arithmetic sequences if possible.
+
+    Given two ``ImageSet``s *a* and *b* representing arithmetic sequences
+    respectively defined by $n \mapsto p n + q$ and $n \mapsto r n + s$, with
+    $p, q, r, s$ complex numbers, ``union_sets()`` checks whether it is
+    possible to reduce the two arithmetic sequences to a single sequence and,
+    if possible, returns a new ``ImageSet`` based on this combined arithmetic
+    sequence.
+
+    No other type of ``ImageSet`` unions is currently supported.
+
+    """
+    if (a.base_set != S.Integers or b.base_set != S.Integers or
+        not all(isinstance(imgset.lamda.expr, Expr) for imgset in (a, b))):
+            return None
+
+    def extract_linear_coeffs(imgset):
+        """Returns $p$, $q$ in $n \mapsto p n + q$; else None."""
+        expr = imgset.lamda.expr
+        var = imgset.lamda.variables
+        if len(var) > 1:
+            return None
+        var = var[0]
+        poly = expr.as_poly(var)
+        if poly and poly.is_linear:
+            return poly.LC(), poly.TC()
+
+    pq = extract_linear_coeffs(a)
+    if not pq:
+        return
+    rs = extract_linear_coeffs(b)
+    if not rs:
+        return
+    p, q, r, s = *pq, *rs
+
+    def in_sequence(x, p, q):
+        # x == p*n + q iff n == (x - q)/p
+        return cancel((x - q)/p).is_integer == True
+    def is_subset(p, q, r, s):
+        # checks whether {p*n + q} is a subset of {r*m + s}
+        if not cancel(p/r).is_integer:
+            return False
+        return in_sequence(q, r, s)
+
+    # 1st possibility: `a` is a subset of `b` or vice-versa
+    if is_subset(p, q, r, s):
+        return b.doit()  # make output a bit more deterministic
```

Comment:
What does the doit do?

Reply:
It canonicalizes the lambda: e.g. `3*n + 5` would be reduced to `3*n + 2`. This also happens at creation with `imageset`. (Both end up calling the same `SetExpr` code.) Without the `doit`  `{3*n + 5} + {3*n+8}` could return either of those two (unmodified) `ImageSet`s depending on the order they're passed to `union_sets`.

In general that's one aspect that I'm a bit unsure about. I noticed that the order in which dispatch calls out to `union_sets` seems to be not deterministic. And this in turn could lead to non-deterministic return values and also non-deterministic test failures...

So I added the `.doit()` since it can't hurt. But `.doit()` only handles some basic situations if I remember it correctly. So for basic lambdas `a*n+b` with `a` and `b` real this should be fine. But I'm not so sure about the case where `a` and `b` are complex. I specifically coded the function to apply to the general case, so that it also takes care of e.g. `{2*n*I*pi} + {2*n*I*pi + I*pi}` as given by hyperbolic trig equations (and it even works for `a` neither real nor pure imaginary, so that the elements actually lie on an oblique line through the complex plane).

I still have to add tests for those cases and then I'll see clearer...

Reply:
Is there a way to compute that more directly from `p, q, r, s`? In the case of `3*n + 8` I can use `8 % 3`.

Reply:
Actually `doit` does a bit more: it also extracts minus signs under some conditions:
https://github.com/sympy/sympy/blob/4cee80af497a25f2c24acbbe20862b54698d5adc/sympy/sets/handlers/functions.py#L178-L212
Also, as soon as complex numbers enter the picture, a simple `Mod` is no longer sufficient. So quite a bit of the above logic would have to be duplicated.